streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -380,43 +372,46 @@ abstract class ZSink[-R, +E, -I, +Z] private (
       ref <- ZRef.make[State[Z, Z1]](BothRunning).toManaged_
       p1  <- self.push
       p2  <- that.push
-      push: Push[R1, E1, I1, Z2] = {
+      push: Push[R1, E1, I1, L1, Z2] = {
         in =>
           ref.get.flatMap {
             state =>
-              val newState: ZIO[R1, Either[E1, Z2], State[Z, Z1]] = {
+              val newState: ZIO[R1, (Either[E1, Z2], Chunk[L1]), State[Z, Z1]] = {
                 state match {
                   case BothRunning => {
-                    val l = p1(in).foldM({
-                      case Left(e)  => ZIO.fail(Left(e))
-                      case Right(z) => ZIO.succeedNow(Some(z))
+                    val l: ZIO[R, (Either[E1, Z2], Chunk[L1]), Option[(Z, Chunk[L])]] = p1(in).foldM({
+                      case (Left(e), l)  => Push.fail(e, l)
+                      case (Right(z), l) => ZIO.succeedNow(Some((z, l)))
                     }, _ => ZIO.succeedNow(None))
-                    val r = p2(in).foldM({
-                      case Left(e)  => ZIO.fail(Left(e))
-                      case Right(z) => ZIO.succeedNow(Some(z))
+                    val r: ZIO[R1, (Left[E1, Nothing], Chunk[L1]), Option[(Z1, Chunk[L1])]] = p2(in).foldM({
+                      case (Left(e), l)  => Push.fail(e, l)
+                      case (Right(z), l) => ZIO.succeedNow(Some((z, l)))
                     }, _ => ZIO.succeedNow(None))
 
                     l.zipPar(r).flatMap {
-                      case (Some(z), Some(z1)) => ZIO.fail(Right(f(z, z1)))
-                      case (Some(z), None)     => ZIO.succeedNow(LeftDone(z))
-                      case (None, Some(z1))    => ZIO.succeedNow(RightDone(z1))
-                      case (None, None)        => ZIO.succeedNow(BothRunning)
+                      case (Some((z, l)), Some((z1, l1))) => {
+                        val minLeftover = if (l.length > l1.length) l1 else l
```

Comment:
What's your reasoning behind using the minimum? I can see arguments for both choices, just wondering what you're thinking.

Reply:
Consider `(s1.zipPar(s2)).zip(s3)`. We have two lineages of consumers: `s1.zip(s3)` and `s2.zip(s3)`. 
Returning the maximum means that one of the lineages consumes some elements twice. 
Which is the most surprising behavior IMO. 
We have consumers that ignore some of the input but no consumer processes an element twice (w/o explicitly being asked for).

Btw, what are the arguments for the opposite?